src/user-service.ts

```diff
@@ -0,0 +1,127 @@
+import { Result } from '@internetarchive/result-type';
+import 'cookie-store';
+import { UserResponse, UserServiceResponse } from './models/response';
+import { User } from './models/user';
+import type { UserInterface } from './models/user';
+import { UserServiceError, UserServiceErrorType } from './user-service-error';
+
+export interface UserServiceInterface {
+  /**
+   * Return the current user info or null if no current user
+   *
+   * @returns {Promise<Result<UserInterface, UserServiceError>>}
+   * @memberof UserServiceInterface
+   */
+  getLoggedInUser(): Promise<Result<UserInterface, UserServiceError>>;
+}
+
+/**
+ * An interface for a cache to conform to for caching User info
+ *
+ * @export
+ * @interface UserServiceCacheInterface
+ */
+export interface UserServiceCacheInterface {
+  set(options: { key: string; value: any; ttl?: number }): Promise<void>;
+
+  get(key: string): Promise<any>;
+}
+
+export class UserService implements UserServiceInterface {
+  private userServiceEndpoint: string;
+
+  private cache?: UserServiceCacheInterface;
+
+  private userCacheKey;
+
+  private cacheTTL?: number;
+
+  constructor(options?: {
+    userServiceEndpoint?: string;
+    cache?: UserServiceCacheInterface;
+    cacheTTL?: number;
+    userCacheKey?: string;
+  }) {
+    this.userServiceEndpoint =
+      options?.userServiceEndpoint ??
+      'https://archive.org/services/user.php?op=whoami';
+    this.cache = options?.cache;
+    this.cacheTTL = options?.cacheTTL;
+    this.userCacheKey = options?.userCacheKey ?? 'loggedInUserInfo';
+  }
+
+  /** @inheritdoc */
+  async getLoggedInUser(): Promise<Result<UserInterface, UserServiceError>> {
+    const hasCookies = await this.hasArchiveOrgLoggedInCookies();
+    if (!hasCookies)
+      return {
+        error: new UserServiceError(UserServiceErrorType.userNotLoggedIn),
+      };
+
+    const persistedUser = await this.getPersistedUser();
+    if (persistedUser) {
+      const user = User.fromUserResponse(persistedUser);
+      return { success: user };
+    }
+
+    let response: Response;
+    try {
+      response = await fetch(this.userServiceEndpoint, {
+        credentials: 'include',
+      });
+    } catch (err) {
+      const message = err instanceof Error ? err.message : err;
+      return {
+        error: new UserServiceError(UserServiceErrorType.networkError, message),
+      };
+    }
+
+    let result: UserServiceResponse;
+    try {
```

Comment:
q: are the use of the try/catches in this (& previous block) to serve the proper error?

what was your main hesitation to not "read" from this original fetch call: `response = await fetch(this.userServiceEndpoint, {...})`

Reply:
Great question!

So my thinking behind this was making the backend somewhat opaque to the consumer. The consumer doesn’t necessarily know that a `fetch` call is getting made inside `UserService` or that a decoding step happens so I didn’t want to have to move that `try / catch` up to the consumer to catch potentially a wide range of errors without knowing the context of the error.

What this does, and where the `Result` type comes into play, is it lets you create specific, type-safe errors so when the consumer of this gets a response back, it can only be a handful of errors, the ones we emit from `UserService`. This lets the consumer switch on the error type and handle as-needed. ie:

```
switch (result.error?.type) {
  case UserServiceErrorType.userNotLoggedIn: <-- you get autocomplete with UserServiceErrorType since it's defined in the `Result` response
    console.info('User not logged in');
    break;
  case UserServiceErrorType.networkError:
    ...
  default:
    ...
```